File: examples/adapters/hpo_optimizer_jsonl.R
#!/usr/bin/env Rscript
# Minimal R optimizer for dag-ml-cli run-host-hpo. DAG-ML owns trial execution,
# fold scoring, pruning and native checkpoints; this host only speaks JSONL.
suppressPackageStartupMessages(library(jsonlite))

reply_for <- function(event) {
  operation <- event$operation
  if (identical(operation, "init")) {
    return(list(prepared_checkpoint = NULL, interrupted = list()))
  }
  if (identical(operation, "ask")) {
    return(list(params = list(n_components = event$trial_index + 1L)))
  }
  if (identical(operation, "report_intermediate")) {
    return(list(prune = FALSE))
  }
  if (operation %in% c("tell", "pruned", "fail", "prepare_terminal", "checkpoint")) {
    return(list(ok = TRUE))
  }
  list(error = paste0("unsupported HPO operation ", operation))
}

input <- file("stdin", open = "r")
repeat {
  line <- readLines(input, n = 1L, warn = FALSE)
  if (length(line) == 0L) break
  event <- fromJSON(line, simplifyVector = FALSE)
  cat(toJSON(reply_for(event), auto_unbox = TRUE, null = "null"), "\n", sep = "")
  flush(stdout())
}
close(input)

File: examples/adapters/hpo_optimizer_jsonl.py
#!/usr/bin/env python3
"""Minimal JSONL host optimizer for ``dag-ml-cli run-host-hpo`` examples.

The scheduler owns trials and scores. This adapter only proposes parameters,
receives fold feedback, and acknowledges durable checkpoint boundaries.
"""

import json
import sys


def main():
    for line in sys.stdin:
        event = json.loads(line)
        operation = event["operation"]
        if operation == "init":
            reply = {"prepared_checkpoint": None, "interrupted": []}
        elif operation == "ask":
            reply = {"params": {"n_components": event["trial_index"] + 1}}
        elif operation == "report_intermediate":
            reply = {"prune": False}
        elif operation in {"tell", "pruned", "fail", "prepare_terminal", "checkpoint"}:
            reply = {"ok": True}
        else:
            reply = {"error": f"unsupported HPO operation {operation}"}
        print(json.dumps(reply), flush=True)


if __name__ == "__main__":
    main()
